<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ sCCIgen</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sCCIgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-get-started" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Get started</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-get-started"><li><a class="dropdown-item" href="../articles/installation.html">Installation</a></li>
    <li><a class="dropdown-item" href="../articles/docker.html">Using sCCIgen Docker image</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Documentation</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/Interface.html">Running a simulation using the interactive interface</a></li>
    <li><a class="dropdown-item" href="../articles/Rpackage_snRNAseq.html">Running a simulation based on snRNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/Rpackage_SRT.html">Running a simulation based on SRT data</a></li>
    <li><a class="dropdown-item" href="../articles/Rpackage_SRT_region.html">Running a simulation based on SRT data region</a></li>
    <li><a class="dropdown-item" href="../articles/Rpackage_unpaired.html">Running a simulation based on unpaired expression and spatial data</a></li>
    <li><a class="dropdown-item" href="../articles/converters.html">Convert sCCIgen to other objects</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/songxiaoyu/sCCIgen/" aria-label="github"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="Add.Distance.Asso.Pattern.html">Add.Distance.Asso.Pattern()</a></code>

        </dt>
        <dd>Add cell-cell expr-distance interaction to a pair of cell types</dd>
      </dl><dl><dt>

          <code><a href="Add.Expr.Asso.Pattern.html">Add.Expr.Asso.Pattern()</a></code>

        </dt>
        <dd>Add cell-cell expr-expr interaction pattern to a pair of cell types</dd>
      </dl><dl><dt>

          <code><a href="Add.Spatial.Expr.Pattern.html">Add.Spatial.Expr.Pattern()</a></code>

        </dt>
        <dd>Adds spatial differential expressed pattern (region-specific effects) to a cell type.</dd>
      </dl><dl><dt>

          <code><a href="Est_ModelPara.html">Est_ModelPara()</a></code>

        </dt>
        <dd>Estimate Marginal distribution and Gaussian Copula for Gene Expression Matrix</dd>
      </dl><dl><dt>

          <code><a href="ExprDistanceTable.html">ExprDistanceTable()</a></code>

        </dt>
        <dd>Save cell expression - distance patterns for simulation</dd>
      </dl><dl><dt>

          <code><a href="ExprExprTable.html">ExprExprTable()</a></code>

        </dt>
        <dd>Save cell expression - cell expression association patterns for simulation</dd>
      </dl><dl><dt>

          <code><a href="MergeRegion.html">MergeRegion()</a></code>

        </dt>
        <dd>Merge spatial and expression data from multiple regions</dd>
      </dl><dl><dt>

          <code><a href="ParaCellsNoST.html">ParaCellsNoST()</a></code>

        </dt>
        <dd>Use parameters to simulate cell location (no existing spatial info)</dd>
      </dl><dl><dt>

          <code><a href="ParaCellsST.html">ParaCellsST()</a></code>

        </dt>
        <dd>Use parameters to simulate cell location based on modeling of existing SRT</dd>
      </dl><dl><dt>

          <code><a href="ParaDigest.html">ParaDigest()</a></code>

        </dt>
        <dd>Digest the parameter file.</dd>
      </dl><dl><dt>

          <code><a href="ParaExistingCellsST.html">ParaExistingCellsST()</a></code>

        </dt>
        <dd>Use parameters to simulate cell location (direct output from existing spatial data)</dd>
      </dl><dl><dt>

          <code><a href="ParaExpr.html">ParaExpr()</a></code>

        </dt>
        <dd>Simualte gene expression data based on parameters</dd>
      </dl><dl><dt>

          <code><a href="ParaSimulation.html">ParaSimulation()</a></code>

        </dt>
        <dd>(Main Function) Simulate spatially resolved transcriptomics data from a parameter file.</dd>
      </dl><dl><dt>

          <code><a href="Pattern.Adj.html">Pattern.Adj()</a></code>

        </dt>
        <dd>Adjust the count data for all cells in all regions based on the input spatial patterns</dd>
      </dl><dl><dt>

          <code><a href="RandomRegionWindow.html">RandomRegionWindow()</a></code>

        </dt>
        <dd>Generate random connected region in a window</dd>
      </dl><dl><dt>

          <code><a href="SpatialTable.html">SpatialTable()</a></code>

        </dt>
        <dd>SpatialTable</dd>
      </dl><dl><dt>

          <code><a href="Use_scDesign2.html">Use_scDesign2()</a></code>

        </dt>
        <dd>Generate expression profile under no spatial patterns.</dd>
      </dl><dl><dt>

          <code><a href="cell.loc.existing.fc.html">cell.loc.existing.fc()</a></code>

        </dt>
        <dd>Generate cell location data by using existing SRT data directly.</dd>
      </dl><dl><dt>

          <code><a href="cell.loc.fc.html">cell.loc.fc()</a></code>

        </dt>
        <dd>Generate cell location data from parameters.</dd>
      </dl><dl><dt>

          <code><a href="cell.loc.model.fc.html">cell.loc.model.fc()</a></code>

        </dt>
        <dd>Generate cell location data by modeling the spatial information of existing data.</dd>
      </dl><dl><dt>

          <code><a href="cellProximityTable.html">cellProximityTable()</a></code>

        </dt>
        <dd>Save cell attraction and inhibition patterns for simulation</dd>
      </dl><dl><dt>

          <code><a href="connectUp.html">connectUp()</a></code>

        </dt>
        <dd>Assign connected regions in a window</dd>
      </dl><dl><dt>

          <code><a href="df_to_config.html">df_to_config()</a></code>

        </dt>
        <dd>Convert a data.frame to config.yml</dd>
      </dl><dl><dt>

          <code><a href="fit_Gaussian_copula.html">fit_Gaussian_copula()</a></code>

        </dt>
        <dd>Fit a Gaussian copula model for a count matrix of a single cell type</dd>
      </dl><dl><dt>

          <code><a href="fit_model_scDesign2.html">fit_model_scDesign2()</a></code>

        </dt>
        <dd>Fit models for a count matrix</dd>
      </dl><dl><dt>

          <code><a href="fit_wo_copula.html">fit_wo_copula()</a></code>

        </dt>
        <dd>Fit a (w/o copula) model for a count matrix of a single cell type</dd>
      </dl><dl><dt>

          <code><a href="loadRData.html">loadRData()</a></code>

        </dt>
        <dd>load RData with new assigned file name</dd>
      </dl><dl><dt>

          <code><a href="multicell.html">multicell()</a></code>

        </dt>
        <dd>Assigns cells into spatial spots for multi-cell resolution data</dd>
      </dl><dl><dt>

          <code><a href="preprocessGiotto.html">preprocessGiotto()</a></code>

        </dt>
        <dd>Process SRT data using Giotto for CCI analysis.</dd>
      </dl><dl><dt>

          <code><a href="run_interactive_sCCIgen.html">run_interactive_sCCIgen()</a></code>

        </dt>
        <dd>Run interactive sCCIgen</dd>
      </dl><dl><dt>

          <code><a href="sCCIgen_to_Giotto.html">sCCIgen_to_Giotto()</a></code>

        </dt>
        <dd>Convert sCCIgen output to a Giotto object</dd>
      </dl><dl><dt>

          <code><a href="sCCIgen_to_Seurat.html">sCCIgen_to_Seurat()</a></code>

        </dt>
        <dd>Convert sCCIgen output to a Seurat object</dd>
      </dl><dl><dt>

          <code><a href="sCCIgen_to_SpatialExperiment.html">sCCIgen_to_SpatialExperiment()</a></code>

        </dt>
        <dd>Convert sCCIgen output to a SpatialExperiment object</dd>
      </dl><dl><dt>

          <code><a href="scDesign2.revised.html">scDesign2.revised()</a></code>

        </dt>
        <dd>Simulate a count matrix for a single cell type</dd>
      </dl><dl><dt>

          <code><a href="simu.window.html">simu.window()</a></code>

        </dt>
        <dd>Simulate the window of spatial data</dd>
      </dl><dl><dt>

          <code><a href="simulate_count_copula.html">simulate_count_copula()</a></code>

        </dt>
        <dd>Simulate a count matrix for a single cell type based on a copula model</dd>
      </dl><dl><dt>

          <code><a href="simulate_count_ind.html">simulate_count_ind()</a></code>

        </dt>
        <dd>Simulate a count matrix for a single cell type based on a (w/o copula model)</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/songxiaoyu/sCCIgen" class="external-link">Xiaoyu Song</a>, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. ChÃ¡vez-Fuentes</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

