<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Running a simulation based on SRT data • sCCIgen</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running a simulation based on SRT data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sCCIgen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-get-started" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Get started</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-get-started">
<li><a class="dropdown-item" href="../articles/installation.html">Installation</a></li>
    <li><a class="dropdown-item" href="../articles/docker.html">Using sCCIgen Docker image</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Documentation</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><a class="dropdown-item" href="../articles/Interface.html">Running a simulation using the interactive interface</a></li>
    <li><a class="dropdown-item" href="../articles/Rpackage_snRNAseq.html">Running a simulation based on snRNAseq data</a></li>
    <li><a class="dropdown-item" href="../articles/Rpackage_SRT.html">Running a simulation based on SRT data</a></li>
    <li><a class="dropdown-item" href="../articles/Rpackage_SRT_region.html">Running a simulation based on SRT data region</a></li>
    <li><a class="dropdown-item" href="../articles/Rpackage_unpaired.html">Running a simulation based on unpaired expression and spatial data</a></li>
    <li><a class="dropdown-item" href="../articles/converters.html">Convert sCCIgen to other objects</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">News</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/songxiaoyu/sCCIgen/" aria-label="github"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">


<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Running a simulation based on SRT data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/songxiaoyu/sCCIgen/blob/HEAD/vignettes/Rpackage_SRT.Rmd" class="external-link"><code>vignettes/Rpackage_SRT.Rmd</code></a></small>
      <div class="d-none name"><code>Rpackage_SRT.Rmd</code></div>
    </div>

    
    
<p>This tutorial will run a simulation with <code>sCCIgen</code> without
the interactive interface based on SRT data.</p>
<div class="section level2" number="1">
<h2 id="load-r-package">
<span class="header-section-number">1</span> Load R package<a class="anchor" aria-label="anchor" href="#load-r-package"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://songxiaoyu.github.io/sCCIgen/">sCCIgen</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2" number="2">
<h2 id="load-and-clean-sample-data">
<span class="header-section-number">2</span> Load and clean sample
data<a class="anchor" aria-label="anchor" href="#load-and-clean-sample-data"></a>
</h2>
<p>Download sample data from <a href="https://github.com/songxiaoyu/sCCIgen_data/tree/main/input_data" class="external-link uri">https://github.com/songxiaoyu/sCCIgen_data/tree/main/input_data</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"SeqFishPlusCortex_2025_expr.Rdata"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"SeqFishPlusCortex_2025_spatial.Rdata"</span><span class="op">)</span></span>
<span><span class="va">anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://drieslab.github.io/GiottoClass/reference/row-plus-colnames-generic.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span></span>
<span><span class="va">expr</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">spatial</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2" number="3">
<h2 id="analysis-of-the-existing-data-to-provide-insights-into-the-parameters-of-the-simulation">
<span class="header-section-number">3</span> Analysis of the
existing data to provide insights into the parameters of the
simulation<a class="anchor" aria-label="anchor" href="#analysis-of-the-existing-data-to-provide-insights-into-the-parameters-of-the-simulation"></a>
</h2>
<p>Users can split the <code>sCCIgen</code> simulation into (1) <u>
model fitting </u> and (2) <u> simulation using fitted model and
user-provided parameters </u> steps to expedite the simulations.</p>
<p>It is especially helpful if the number of genes and/or cells are very
large and users want to run simulation for more than once. By splitting
the simulation into these two steps, users can estimate model parameters
only once and save the results for multiple use.</p>
<div class="section level3" number="3.1">
<h3 id="task-1-estimate-model-parameters-from-the-snrnaseq-for-simulation">
<span class="header-section-number">3.1</span> Task 1: Estimate
model parameters from the snRNAseq for simulation<a class="anchor" aria-label="anchor" href="#task-1-estimate-model-parameters-from-the-snrnaseq-for-simulation"></a>
</h3>
<p>This is part is to fit the expression data. When
sim_method==“copula”, it will fit both the gene marginal distribution
and gene-gene correlation. When sim_method==“ind”, it will only fit the
gene marginal distribution.</p>
<p>Note: If the number of genes or cells are large, model fitting may
take some time. It is suggested to select a reasonable sample size
(e.g. &lt;10000 per cell type) before the model fitting, as more cells
may not be needed improve the estimation. Similarly, if some genes are
extremely zero-inflated, narrowing the simulation to reasonbly variable
genes is an option.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># model fitting w/o gene-gene correlation (fast but less accurate estimation)</span></span>
<span><span class="va">ModelEst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Est_ModelPara.html">Est_ModelPara</a></span><span class="op">(</span>expr <span class="op">=</span> <span class="va">expr</span>, </span>
<span>                          anno <span class="op">=</span> <span class="va">anno</span>, </span>
<span>                          sim_method <span class="op">=</span> <span class="st">"ind"</span>,</span>
<span>                          ncores <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">ModelEst</span>, file <span class="op">=</span> <span class="st">"SeqFishPlusCortex_fit_wo_cor.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># model fitting with gene-gene correlation</span></span>
<span><span class="va">ModelEst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Est_ModelPara.html">Est_ModelPara</a></span><span class="op">(</span>expr <span class="op">=</span> <span class="va">expr</span>, </span>
<span>                          anno <span class="op">=</span> <span class="va">anno</span>, </span>
<span>                          sim_method <span class="op">=</span> <span class="st">"copula"</span>, </span>
<span>                          ncores <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">ModelEst</span>, file <span class="op">=</span> <span class="st">"SeqFishPlusCortex_fit_w_cor.RDS"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3" number="3.2">
<h3 id="task-2-estimate-ccis-in-the-input-data-following-the-existing-giotto-pipeline">
<span class="header-section-number">3.2</span> Task 2: Estimate CCIs
in the input data following the existing Giotto pipeline<a class="anchor" aria-label="anchor" href="#task-2-estimate-ccis-in-the-input-data-following-the-existing-giotto-pipeline"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://drieslab.github.io/Giotto/" class="external-link">Giotto</a></span><span class="op">)</span></span>
<span><span class="co"># Preprocess data with Giotto pipeline</span></span>
<span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocessGiotto.html">preprocessGiotto</a></span><span class="op">(</span>expr_data <span class="op">=</span> <span class="va">expr</span>, </span>
<span>                       spatial_data <span class="op">=</span> <span class="va">spatial</span>, </span>
<span>                       run_hvg <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       run_Dist_network <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       dis.cut<span class="op">=</span><span class="fl">150</span><span class="op">)</span></span>
<span><span class="va">db</span></span>
<span><span class="co"># Estimate cell-cell attraction and inhibition patterns, and save in pre-defined folder</span></span>
<span><span class="fu"><a href="../reference/cellProximityTable.html">cellProximityTable</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">db</span>, </span>
<span>                   abs_enrichm <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   p_adj <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                   spatial_network_name <span class="op">=</span> <span class="st">"distance_based_network"</span>,</span>
<span>                   save.unfiltered<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                   output_file <span class="op">=</span> <span class="st">"est_CCI_dist_dist.csv"</span>,</span>
<span>                   seed<span class="op">=</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Estimate gene expressions of cells impacted by their neighbors, and save in pre-defined folder</span></span>
<span><span class="fu"><a href="../reference/ExprDistanceTable.html">ExprDistanceTable</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">db</span>, </span>
<span>                  in_hvg <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                  region_specific <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                  spatial_network_name <span class="op">=</span> <span class="st">"distance_based_network"</span>,</span>
<span>                  abs_log2fc_ICG <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                  p_adj <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                  output_file <span class="op">=</span> <span class="st">"est_CCI_dist_expr.csv"</span>,</span>
<span>                  seed<span class="op">=</span><span class="fl">123</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Estimate gene expressions of cells impacted by gene expressions of neighboring cells, narrow to</span></span>
<span><span class="co">#  known pairs, such as ligand and receptor pairs, and save in pre-defined folder</span></span>
<span><span class="fu"><a href="../reference/ExprExprTable.html">ExprExprTable</a></span><span class="op">(</span>gobject <span class="op">=</span> <span class="va">db</span>, </span>
<span>              database <span class="op">=</span> <span class="st">"mouse"</span>, </span>
<span>              region_specific <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>              spatial_network_name <span class="op">=</span> <span class="st">"distance_based_network"</span>,</span>
<span>              p_adj <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>              abs_log2fc_LR <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>              output_file <span class="op">=</span> <span class="st">"est_CCI_expr_expr.csv"</span>,</span>
<span>              seed<span class="op">=</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2" number="4">
<h2 id="create-a-parameter-file">
<span class="header-section-number">4</span> Create a parameter
file<a class="anchor" aria-label="anchor" href="#create-a-parameter-file"></a>
</h2>
<p>Users need to create a parameter file. The sample parameter file for
snRNAseq based simulation is <a href="https://github.com/songxiaoyu/sCCIgen_data/tree/main/sample_parameter_file/SRT" class="external-link">here</a>
for downloading and filling in to perform simulations.</p>
</div>
<div class="section level2" number="5">
<h2 id="perform-the-entire-simulation-and-save-the-results-">
<span class="header-section-number">5</span> Perform the entire
simulation and save the results.<a class="anchor" aria-label="anchor" href="#perform-the-entire-simulation-and-save-the-results-"></a>
</h2>
<p>Assuming you already have a parameter file, you can run the entire
simulation using codes like this:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate default data - using existing cells but simulate expression with ground truth</span></span>
<span><span class="va">input</span> <span class="op">&lt;-</span> <span class="st">"SeqFishPlus_default.tsv"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/ParaSimulation.html">ParaSimulation</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">input</span>, ModelFitFile <span class="op">=</span> <span class="va">model_param_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate new cells and genes where genes have region specific expressions</span></span>
<span><span class="va">input</span> <span class="op">&lt;-</span> <span class="st">"SeqFishPlus_RegionDiffGenes.tsv"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/ParaSimulation.html">ParaSimulation</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">input</span>, ModelFitFile <span class="op">=</span> <span class="va">model_param_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate new cells and genes where expression of genes are associated with distances to other cells.</span></span>
<span><span class="va">input</span> <span class="op">&lt;-</span> <span class="st">"SeqFishPlus_ICGs.tsv"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/ParaSimulation.html">ParaSimulation</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">input</span>, ModelFitFile <span class="op">=</span> <span class="va">model_param_path</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate new cells and genes where expression of genes are associated with distances to other cells.</span></span>
<span><span class="va">input</span> <span class="op">&lt;-</span> <span class="st">"SeqFishPlus_LR.tsv"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/ParaSimulation.html">ParaSimulation</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">input</span>, ModelFitFile <span class="op">=</span> <span class="va">model_param_path</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/songxiaoyu/sCCIgen" class="external-link">Xiaoyu Song</a>, <a href="https://josschavezf.github.io" class="external-link">Joselyn C. Chávez-Fuentes</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
